### PR TITLE
[NEW] HF Sync plugin

### PR DESCRIPTION
**Pull request by frascuchon**
With this plugin, users can sync data from/to the HF datasets hub and an Argilla instance.

Refs https://github.com/argilla-io/argilla/issues/2614  

## Description

Since this plugin has several optional configurations, can cover several use cases. Let's enumerate some of them

1. Automatically import an HF dataset into your Argilla instance (by setting the `hf_source`)
2. Export total or partial dataset records into an HF dataset (by setting the `hf_target` and `rg_dataset` parameters)
3. Sync HF datasets with changes from Argilla (by setting the `hf_source=hf_target` and `rg_query=None`). Be careful with this since the dataset will be exported totally each `hf_push_to_hub_frequency` seconds.

### CODE REVIEW

File: argilla_plugins/datasets/hf_sync.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+@dataclasses.dataclass
+class Config:
+
+    hf_source: Optional[str] = None
+    hf_target: Optional[str] = None
+    hf_token: Optional[str] = None
+    hf_split: str = "train"
+    hf_push_to_hub_frequency: int = 60
+
+    rg_task: TaskType = TaskType.text_classification
```
Comment (davidberenstein1957) [inline argilla_plugins/datasets/hf_sync.py:23]: Why do we default to `text_classification`?

File: argilla_plugins/datasets/hf_sync.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+from argilla.client.sdk.commons.errors import NotFoundApiError
+from argilla.server.commons.models import TaskType
+
+
+@dataclasses.dataclass
+class Config:
+
+    hf_source: Optional[str] = None
+    hf_target: Optional[str] = None
+    hf_token: Optional[str] = None
```
Comment (davidberenstein1957) [inline argilla_plugins/datasets/hf_sync.py:19]: maybe we should be able to set this programmatically via env vars instead of only CLI?

**Comment by dvsrepo** [comment]
Looks good!!

One question: 

This looks like a two-way sync like the one we have for Alpaca (exactly what we want for Spaces)

For more general use cases,  could/do we also cover the use case where a user just wants to make sure the dataset is stored in a Hub Dataset periodically and not loading the dataset from the Hub back to the Argilla dataset?

(Note: I only skimmed through the code)

File: argilla_plugins/datasets/hf_sync.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+import argilla as rg
+from argilla.client.sdk.commons.errors import NotFoundApiError
+from argilla.server.commons.models import TaskType
+
+
+@dataclasses.dataclass
+class Config:
+
+    hf_source: Optional[str] = None
```
Comment (davidberenstein1957) [inline argilla_plugins/datasets/hf_sync.py:17]: adding the source too makes the use cases really flexible.

**Review by davidberenstein1957 — COMMENTED** [review]
Thanks for the PR paco! It looks great and I think I can use a lot of the logic w.r.t. threading to formalize the other plugins. I have some small comments, and I think we should add documentation on how to use it.  Additionallt, Jose from Datamaran was asking about something similar to this, so he might be interested.

File: argilla_plugins/datasets/hf_sync.py (hunk 1 above)
Comment (frascuchon) [reply argilla_plugins/datasets/hf_sync.py:23]: Not sure. Yes, maybe can be a required value just to force users to which kind of dataset is expecting to sync.

File: argilla_plugins/datasets/hf_sync.py (hunk 2 above)
Comment (frascuchon) [reply argilla_plugins/datasets/hf_sync.py:19]: Yes. Maybe the Config class can be a pedantic `BaseSettings` class that gives automatic support with env variables.

**Comment by davidberenstein1957** [comment]
Additionally, if the code is not robust enough, you could try to setup a retry-backoff with smaller chunk sizes?Cheers,David On 30 Mar 2023, at 18:43, Francisco Aranda ***@***.***> wrote:﻿
I will try to take a look tomorrow with your feedback.

**Comment by frascuchon** [comment]
> Additionally, if the code is not robust enough, you could try to setup a retry-backoff with smaller chunk sizes?Cheers,David On 30 Mar 2023, at 18:43, Francisco Aranda ***@***.***> wrote:﻿ I will try to take a look tomorrow with your feedback. —Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because your review was requested.Message ID: ***@***.***>

Great! But I think this behavior should be in the core rg.log/rg.load methods

**Comment by frascuchon** [comment]
The plugin is working now by using environment variables. The name of variables is aligned with the naming of attributes in HuggingfaceSyncConfig class.

I've created a space [here](https://huggingface.co/spaces/frascuchon/argilla-space/blob/main/Dockerfile). 

As TODO that can be tackled in another PR:

- Add retries importing data when argilla is not ready yet.
- Allow restarting the plugin. This is not possible for now. A new plugin instance must be created to achieve it.

Also, it would be great to package the argilla-plugins package with the quickstart image.

**Comment by dvsrepo** [comment]
ok! 

what's left to close this version @frascuchon @davidberenstein1957 ?

**Comment by dvsrepo** [comment]
Maybe a brief description of how to use it?

**Review by davidberenstein1957 — APPROVED** [review]

**Pull request closed without merging**